packages/headless/src/api/service/case-assist/case-assist-api-client.ts

```diff
@@ -0,0 +1,111 @@
+import {Logger} from 'pino';
+import {
+  NoopPreprocessRequestMiddleware,
+  PlatformClient,
+} from '../../platform-client';
+import {NoopPreprocessRequest} from '../../preprocess-request';
+import {buildClassifyRequest, ClassifyParam} from './classify/classify-request';
+import {ClassifySuccessContent} from './classify/classify-response';
+import {
+  buildSuggestDocumentsRequest,
+  SuggestDocumentsParam,
+} from './suggest-documents/suggest-documents-request';
+import {SuggestDocumentsSuccessContent} from './suggest-documents/suggest-documents-response';
+
+/**
+ * Initialization options for the `CaseAssistAPIClient`.
+ */
+export interface CaseAssistAPIClientOptions {
+  logger: Logger;
+  renewAccessToken: () => Promise<string>;
+}
+
+/**
+ * Defines a Case Assist API response. It can represent an error or a successful response.
+ */
+export type CaseAssistAPIResponse<TSuccessContent> =
+  | CaseAssistAPISuccessResponse<TSuccessContent>
+  | CaseAssistAPIErrorResponse;
+
+/**
+ * Defines a Case Assist API successful response.
+ */
+export interface CaseAssistAPISuccessResponse<TContent> {
+  success: TContent;
+}
+
+/**
+ * Defines the content of a Case Assist API error response.
+ */
+export interface CaseAssistAPIErrorStatusResponse {
+  statusCode: number;
+  message: string;
+  type: string;
+}
+
+/**
+ * Defines a Case Assist API error response.
+ */
+export interface CaseAssistAPIErrorResponse {
+  error: CaseAssistAPIErrorStatusResponse;
+}
+
+/**
+ * The client to use to interface with the Case Assist API.
+ */
+export class CaseAssistAPIClient {
+  private defaultClientHooks = {
+    preprocessRequest: NoopPreprocessRequest,
```

Comment:
I wasn't sure how to handle this hook. It should probably move to the `CaseAssistAPIClientOptions`. But I guess it might not work when search and case assist are used simultaneously in the engine.

What is a typical use of `preprocessRequest`?

Reply:
That's a hook that is used to allow users to tap "deep" into the network request before it is sent.

For example, we have customers that want to go through their own proxy, which requires them injecting headers that Headless would have no knowledge about otherwise. So they use this hook to tweak the HTTP request to do whatever is needed on their end.

Another use case is to use this as the ultimate "escape hatch" for any new parameter that might have been exposed by the API, but not yet supported by a specific version of headless. In that case, a customer might use preprocess/postprocess hook to modify the request and add those new parameter without needing them to wait for a new version being available/installed etc.

It's used relatively often/is important to support from our experience.

Reply:
Good to know! I'll move it into the client options then. Thanks for the example!